src/extension.ts

```diff
@@ -1,11 +1,115 @@
+    context.subscriptions.push(vscode.languages.registerCompletionItemProvider('csound-csd', {
+        async provideCompletionItems(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken, context: vscode.CompletionContext) {
+            const config = vscode.workspace.getConfiguration("csound");
+            const showHints = config.get("showHints") as boolean;
+            const filePath = vscode.Uri.joinPath(extensionUri, 'src', 'opcodes.json');
+            // Read the file asynchronously and await the result
+            const fileData = await vscode.workspace.fs.readFile(filePath);
+            const completions: vscode.CompletionItem[] = [];
+            if(!showHints){
+                return completions;
+            }
+            // Convert the Buffer (Uint8Array) to a string
+            const jsonString = Buffer.from(fileData).toString('utf-8');
+            const json = JSON.parse(jsonString);
+            json.opcodes.forEach((categoryObj: any) => {
+                categoryObj.opcodes.forEach((opcodeObj: any) => {
+                    // Create a new CompletionItem
+                    let completionItem = new vscode.CompletionItem(opcodeObj.opcodeName, vscode.CompletionItemKind.Snippet);
+
+                    // Set the label to be the opcode name
+                    completionItem.label = opcodeObj.opcodeName;
+
+                    // Forego detail here, as it is not necessary for the opcode list
+                    completionItem.detail = "";
+
+                    const formattedSynopsis = formatSynopsisForMarkdown(`**${opcodeObj.opcodeName}**: *${opcodeObj.description}* \n\n` + opcodeObj.synopsis + '\n\n' + opcodeObj.functionalSynopsis);
+                    completionItem.documentation = new vscode.MarkdownString(formattedSynopsis);
+
+                    // Add the item to the completion list
+                    completions.push(completionItem);
+                });
+            });
+
+
+            return completions; 1
+        }
+    }, ''));
+
+    const disposable = vscode.commands.registerCommand('extension.showOpcodeReference', async () => {
+        const editor = vscode.window.activeTextEditor;
+        if (!editor) {
+            vscode.window.showInformationMessage('No active editor found.');
+            return;
+        }
 
+        const position = editor.selection.active;
+        const wordRange = editor.document.getWordRangeAtPosition(position);
+        const opcode = editor.document.getText(wordRange).trim(); // Get the opcode
+
+        console.log(`Opcode: ${opcode}`); // Log the opcode for debugging
+
+        const config = vscode.workspace.getConfiguration('yourExtension');
+        const customPath = config.get<string>('htmlFilePath') === "Use default location" ? "" : config.get<string>('htmlFilePath');
+        const osPlatform = process.platform;
+        let defaultPath: string;
+
+        if (!customPath) {
+            if (osPlatform === 'win32') {
```

Comment:
Is the `CSDOCDIR` useful here or is it not usually set?  https://csound.com/manual/CommandEnvironment.html